pageserver/src/tenant/storage_layer/layer.rs

```diff
@@ -0,0 +1,1150 @@
+                tokio::time::sleep(backoff).await;
+
+                Err(DownloadError::DownloadFailed)
+            }
+            Err(_gone) => Err(DownloadError::DownloadCancelled),
+        }
+    }
+
+    async fn needs_download(&self) -> Result<Option<NeedsDownload>, std::io::Error> {
+        match tokio::fs::metadata(&self.path).await {
+            Ok(m) => Ok(self.is_file_present_and_good_size(&m).err()),
+            Err(e) if e.kind() == std::io::ErrorKind::NotFound => Ok(Some(NeedsDownload::NotFound)),
+            Err(e) => Err(e),
+        }
+    }
+
+    fn needs_download_blocking(&self) -> Result<Option<NeedsDownload>, std::io::Error> {
+        match self.path.metadata() {
+            Ok(m) => Ok(self.is_file_present_and_good_size(&m).err()),
+            Err(e) if e.kind() == std::io::ErrorKind::NotFound => Ok(Some(NeedsDownload::NotFound)),
+            Err(e) => Err(e),
+        }
+    }
+
+    fn is_file_present_and_good_size(&self, m: &std::fs::Metadata) -> Result<(), NeedsDownload> {
+        // in future, this should include sha2-256 of the file.
+        if !m.is_file() {
+            Err(NeedsDownload::NotFile)
+        } else if m.len() != self.desc.file_size {
+            Err(NeedsDownload::WrongSize {
+                actual: m.len(),
+                expected: self.desc.file_size,
+            })
+        } else {
+            Ok(())
+        }
+    }
+
+    fn info(&self, reset: LayerAccessStatsReset) -> HistoricLayerInfo {
+        let layer_file_name = self.desc.filename().file_name();
+
+        let remote = self
+            .needs_download_blocking()
+            .map(|maybe| maybe.is_some())
+            .unwrap_or(true);
+
+        let access_stats = self.access_stats.as_api_model(reset);
+
+        if self.desc.is_delta {
+            let lsn_range = &self.desc.lsn_range;
+
+            HistoricLayerInfo::Delta {
+                layer_file_name,
+                layer_file_size: self.desc.file_size,
+                lsn_start: lsn_range.start,
+                lsn_end: lsn_range.end,
+                remote,
+                access_stats,
+            }
+        } else {
+            let lsn = self.desc.image_layer_lsn();
+
+            HistoricLayerInfo::Image {
+                layer_file_name,
+                layer_file_size: self.desc.file_size,
+                lsn_start: lsn,
+                remote,
+                access_stats,
+            }
+        }
+    }
+
+    /// Our resident layer has been dropped, we might hold the lock elsewhere.
+    fn on_drop(self: Arc<LayerInner>) {
+        let gc = self.wanted_garbage_collected.load(Ordering::Acquire);
+        let evict = self.wanted_evicted.load(Ordering::Acquire);
+        let can_evict = self.have_remote_client;
+
+        if gc {
+            // do nothing now, only when the whole layer is dropped. gc will end up deleting the
+            // whole layer, in case there is no reference cycle.
```

Comment:
"the whole layer" is ambiguous here

Also, "in case there is no reference cycle" makes it sound like reference cycles are an expected condition; but they aren't. 

```suggestion
            // the deletion work happens in `LayerInner::drop`
```

Reply:
8a27e5889 together with renamed this is quite obvious now.